README.md
# winerva


**Helping to bring the management of stand-alone MS Windows machines under your control!**


## TOC

* [Overview](#overview)
    * [What is this?](#what-is-this)
    * [Why Ansible?](#why-ansible)
    * [Which operating systems?](#which-operating-systems)
    * [What does it do?](#what-does-it-do)
    * [Why the local Administrator account?](#why-the-local-administrator-account)
    * [How do I use it?](#how-do-i-use-it)
* [Configuration](#configuration)


## Overview

### What is this?

This project contains a PowerShell script that prepares a Windows machine to be remotely managed by Ansible, along with a batch file that allows for "one click" operation.


### Why Ansible?

[Ansible](https://www.ansible.com/overview/it-automation) is an open-source computer program that enables remote configuration and application deployment to target machines, without requiring the installation and maintenance of remote agent software. Desired state is written using [YAML](https://en.wikipedia.org/wiki/YAML) and [Jinja2](http://jinja.pocoo.org), which are converted into their matching PowerShell commands and executed remotely via WinRM.

Ansible's combination of free-to-use, agentless architecture, along with a data format that is both easy to read and write, make this an ideal system for low-cost IT maintenance. And, because Ansible communicates with Windows machines via WinRM, it is perfect for environments that contain a number of networked PCs running Windows, but do not have a domain controller.


### Which operating systems?

The list of currently supported operating systems are:

* MS Windows 10 (as of v1.0.0)
* MS Windows 8.1 (as of v1.1.0)
* MS Windows 7 SP1 (as of v.1.2.0)


### What does it do?

The script carries out the following tasks:

* Ensures [Chocolatey](https://chocolatey.org) is installed.
* Upgrades .NET to version 4.5 (latest available patch).
* Upgrades PowerShell to the latest available version.
* Ensures that [Carbon](http://get-carbon.org) for PowerShell is installed.
* Ensures the specified user account is present.
* Ensures that that user account is a member of the local `Administrators` group.
* Adds (if missing) the computer's name to an inventory file, which can later be used by Ansible.


### Why the local Administrator account?

Ansible will use this account when connecting to each target machine. Administrator-level access is required for a lot of configuration actions and application deployments, so it just makes sense. That being said, the credentials should be kept in a safe place (meaning, **not** copied to the target machine), and the password should be [lengthy](https://stormpath.com/blog/5-myths-password-security).


### How do I use it?

1. Download the contents of the `src/` folder to removable media (such as a USB flash drive).
2. Ensure that the contents of the configuration file are correct for the target machine. _You'll need to change the credentials_; if you like, you can use an existing account. The script will automatically add it to the local Administrators group, if it isn't there already.
3. Attach the removable media to the target machine, and execute `run.cmd`. This will [automatically launch the script](http://blog.danskingdom.com/allow-others-to-run-your-powershell-scripts-from-a-batch-file-they-will-love-you-for-it/) in an elevated shell, and bypass the current execution policy setting, in case it is too restrictive.

**Note #1: Depending on which Windows components are already present on the target machine, one or more reboots may be required during the installation process. After each reboot, repeat step #3 above until you see the message "INSTALLATION COMPLETE" in the console window.**

**Note #2: This is script is safe to run more than once on the same machine.**


## Configuration

The following configuration options are stored in the file `config.psd1`, which is in the same location as `run.cmd`. This file is a [PowerShell data file](http://www.dexterposh.com/2017/06/powershell-psd1-files-for-env-config-data.html), comprised of a single hash table. The syntax and grammar are the same as that for PowerShell script files.


### Local administrator account

The credentials for the local administrator account are kept in the file `config.psd1` under the entry `LocalAdmin`. Note that the password must be entered as a plain text value. Refer to the comments within the configuration file for further instructions.


## License

MIT (please refer to the file named LICENSE).


CHANGELOG.md
# Changelog for winerva

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.2.1] - 2018-10-18

### Fixed
* The script no longer encounters an error when `hostnames` is empty.
* Applied style guidelines as per https://poshcode.gitbooks.io/powershell-practice-and-style/.
* Script can be successfully run inside of PowerShell ISE without throwing unhandled exceptions.


## [1.2.0] - 2018-08-28

### Added
* Support for MS Windows 7 SP1. Note this may require one or more restarts.
* Script now installs .NET 4.5
* Script displays verbiage to indicate when the installation process is complete.
* Script includes version number.

### Fixed
* When removing downloaded files, script no longer throws an exception.


## [1.1.0] - 2018-08-24

### Added
* Support for MS Windows 8.1
* Script now installs latest available version of PowerShell.

### Fixed
* When updating `hostnames`, a new line is added between names.
* When updating `hostnames`, empty lines are removed.
* When an exception is thrown, the `catch` block correctly displays the error information.


## [1.0.0] - 2018-08-23

First release.

### Added
* Support for MS Windows 10.


src/winerva.ps1
#Requires -Version 2

<#
.SYNOPSIS
    A PowerShell script that creates a local administrator account and enables
    WinRM, for remote administration via Ansible.

.NOTES
    PREREQUISITES
     * This script is only intended to be run within PowerShell on MS Windows
       machines (not cross-platform compatible).
     * This script can only be executed on MS Windows 7 SP1 or later.

    STYLE
     * This code adheres to the style guidelines published at
       https://poshcode.gitbooks.io/powershell-practice-and-style/

    LICENSE
    Copyright (c) 2018 David Passarelli <dpassarelli@camelotcg.com>

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:
    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.
#>

[CmdletBinding()]
param()

# https://blogs.technet.microsoft.com/heyscriptingguy/2014/12/03/enforce-better-script-practices-by-using-set-strictmode/
#
Set-StrictMode -Version Latest

###################################
##### DEFINE GLOBAL CONSTANTS #####
###################################

# The version of this script. This number follows the specification set by
# https://semver.org
#
# See CHANGELOG.MD for change history.
#
Set-Variable SCRIPT_VERSION -Option Constant -Value "1.2.1"

# The official "version" number reported by Windows 7 with Service Pack 1
# installed. Obtained from
# https://en.wikipedia.org/wiki/List_of_Microsoft_Windows_versions
#
Set-Variable WIN7SP1_VERSION_ID -Option Constant -Value "6.1.7601"

# The value of $ErrorActionPreference before it is changed by this script. It
# will be reset just before the script exits.
#
Set-Variable PREVIOUS_EAP -Option Constant -Value $ErrorActionPreference

# `$PSScriptRoot` was only added as a built-in special variable starting with PS
# version 3.
#
try {
    if ([System.String]::IsNullOrEmpty($PSScriptRoot)) {
        # Adapted from https://stackoverflow.com/a/3667376
        Set-Variable PSScriptRoot -Option ReadOnly -Value (Split-Path $MyInvocation.MyCommand.Path -Parent)
    }
}
catch {
    Set-Variable PSScriptRoot -Option ReadOnly -Value (Split-Path $MyInvocation.MyCommand.Path -Parent)
}

# This hashtable contains various pieces of system information used throughout
# the script.
#
Set-Variable SCRIPT_DATA -Option Constant -Value @{
    # Self-explanatory. The hostname is added to the inventory file for
    # future use by Ansible playbooks.
    #
    HOSTNAME = $Env:ComputerName

    # The location on disk where downloaded scripts will be saved.
    # When the script is about to exit, these will be deleted.
    #
    DOWNLOAD_TO = (Join-Path $Home "Downloads")

    # The currently installed version of Windows.
    #
    WIN_VERSION = (Get-WmiObject -Class Win32_OperatingSystem).Version

    # The latest version of .NET that appears to be installed on this
    # machine. Based on https://stackoverflow.com/a/1565454
    #
    NET_VERSION = (Get-ChildItem -Path 'HKLM:\SOFTWARE\Microsoft\NET Framework Setup\NDP' -Recurse | `
                   Get-ItemProperty -Name Version,Release -ErrorAction SilentlyContinue | `
                   ? { $_.PSChildName -Match '^(?!S)\p{L}'} | `
                   Sort-Object -Property Version | `
                   Select-Object -Last 1).Version

    # The major version number for PowerShell.
    #
    PS_MAJOR_VERSION = $PSVersionTable.PSVersion.Major

    # The absolute path to the inventory file (which contains all of the
    # host names).
    #
    INVENTORY_FILE = (Join-Path $PSScriptRoot "hostnames")
}

###################################
#####     DEFINE FUNCTIONS    #####
###################################

function Write-LogMessage {
    <#
    .SYNOPSIS
        Displays a message on the console. Encapsulating this in a function
        provides an easy way to ensure that debug messages will stand out from
        other output.

    .PARAMETER Msg
        The message to display.
    #>
    [CmdletBinding()]
    param(
        [Parameter(Mandatory = $true)]
        [String]$Msg
    )

    Write-Host "[LOG]  $Msg"
}

function Verify-MinimumWindowsVersion {
    <#
    .SYNOPSIS
        Checks to see if the specified version number meets the minimum
        requirement for this script.

    .PARAMETER Version
        The version identifier for the current OS. Can be determined via
        (Get-WmiObject -Class Win32_OperatingSystem).Version

    .NOTES
        * Returns a boolean value.
        * See https://en.wikipedia.org/wiki/List_of_Microsoft_Windows_versions
    #>
    [CmdletBinding()]
    param(
        [Parameter(Mandatory = $true)]
        [String]$Version
    )

    [Array]$versionParts = $Version.split(".")

    if ([Int]$versionParts[0] -gt 6) {
        return $true
    }

    if ([Int]$versionParts[0] -eq 6) {
        if ([Int]$versionParts[1] -gt 1) {
            return $true
        }

        if ([Int]$versionParts[1] -eq 1 -and [Int]$versionParts[2] -ge 7601) {
            return $true
        }
    }

    $false
}

function Verify-ElevatedShell {
    <#
    .SYNOPSIS
        Checks to see if the current execution context has Administrative
        privileges.

    .NOTES
        * Returns a boolean value.
        * Adapted from https://blogs.msdn.microsoft.com/virtual_pc_guy/2010/09/23/a-self-elevating-powershell-script/
    #>
    [CmdletBinding()]
    param()

    $CurrentUser = New-Object System.Security.Principal.WindowsPrincipal([System.Security.Principal.WindowsIdentity]::GetCurrent())
    Return $CurrentUser.IsInRole([System.Security.Principal.WindowsBuiltInRole]::Administrator)
}

function Download-File {
    <#
    .SYNOPSIS
        Downloads the requested URL, and saves the file to the specified
        location.

    .PARAMETER Url
        The resource to download.

    .PARAMETER Destination
        The local path (including filename) that the download will be saved to.
    #>
    [CmdletBinding()]
    param(
        [Parameter(Mandatory = $true)]
        [String]$Url,

        [Parameter(Mandatory = $true)]
        [String]$Destination
    )

    (New-Object System.Net.WebClient).DownloadFile($Url, $Destination)
}

function Restart-Computer {
    <#
    .SYNOPSIS
        Initiates a restart sequence.
    #>
    [CmdletBinding()]
    param()

    Write-LogMessage "********* MACHINE WILL NOW REBOOT IN 30 SECONDS *********"
    Write-LogMessage "Re-run this script after the machine has come back up."
    shutdown /r /t 30
}

function PrepareFor-Exit {
    <#
    .SYNOPSIS
        Performs general clean-up immediately preceding the end of this script.

    .PARAMETER DownloadLocation
        The location where any files downloaded by this script were saved.

    .PARAMETER PreviousSettingForEAP
        The previous value for `$ErrorActionPreference`.
    #>
    [CmdletBinding()]
    param(
        [Parameter(Mandatory = $true)]
        [String]$DownloadLocation,

        [Parameter(Mandatory = $true)]
        [String]$PreviousSettingForEAP
    )

    ##### CLEAN UP
    Write-LogMessage "Removing downloaded scripts..."
    Get-ChildItem -Path $DownloadLocation -Include "__winerva__*.ps1" -Recurse | % { Remove-Item $_.Fullname }
    Write-LogMessage "done!"

    Write-LogMessage "Restoring previous value of ErrorActionPreference..."
    $ErrorActionPreference = $PreviousSettingForEAP
    Write-LogMessage "...done!"

    Write-LogMessage "END OF LINE"

    # If this script was launched via a "double-click", then keep the console
    # window open so that the user has a chance to review any output before it
    # suddenly disappears. Adapted from
    # http://blog.danskingdom.com/allow-others-to-run-your-powershell-scripts-from-a-batch-file-they-will-love-you-for-it/
    #
    Write-Host "Press any key to exit..."

    try {
        $Host.UI.RawUI.FlushInputBuffer()
        $Host.UI.RawUI.ReadKey("NoEcho,IncludeKeyUp") > $null
    }
    catch {
        # ignore
    }
}

###################################
#####     MAIN ALGORITHM      #####
###################################

##### SET ERRORACTIONPREFERENCE

# The reason for setting this value is to ensure that all unhandled
# exceptions cause the script to throw a terminating error (instead of
# non-terminating).
#
Write-LogMessage "Setting ErrorActionPreference to 'Stop'..."
$ErrorActionPreference = "Stop"
Write-LogMessage "...done!"

try {
    Write-LogMessage "++++++ BEGIN INSTALLATION ++++++"
    Write-LogMessage "++++++++ WINERVA v$SCRIPT_VERSION ++++++++"

    # Echo the collected data back out to the console. This should help with
    # debugging problems in the field.
    #
    Write-LogMessage "Printing script data..."
    Format-List -InputObject $SCRIPT_DATA
    Write-LogMessage "...done!"

    ##### CHECK PREREQUISITES

    Write-LogMessage "Checking prerequisites..."

    if (Verify-ElevatedShell) {
        Write-LogMessage "  * running in elevated shell"
    }
    else {
        throw "This script requires administrative privileges!"
    }

    if (Verify-MinimumWindowsVersion $SCRIPT_DATA.WIN_VERSION) {
        Write-LogMessage "  * OS is at least Windows 7 SP1"
    }
    else {
        throw "This script can only be run on Windows 7 SP1 or later."
    }

    Write-LogMessage "...done!"

    ##### INSTALL CHOCOLATEY

    Write-LogMessage "Checking for Chocolatey..."
    [Boolean]$upgradeChocolatey = $false

    try {
        Get-Command "choco.exe" | Out-Null
        Write-LogMessage "...found."
        $upgradeChocolatey = $true
    }
    catch {
        Write-LogMessage "...not found. Downloading installation script..."

        [String]$pathToChocolateyScript = (Join-Path $SCRIPT_DATA.DOWNLOAD_TO "__winerva__InstallChocolatey.ps1")
        Download-File -Url "https://chocolatey.org/install.ps1" -Dest $pathToChocolateyScript

        Write-LogMessage "...done!"

        Write-LogMessage "Executing..."
        . $pathToChocolateyScript
        Write-LogMessage "...done!"

        # On Windows 7, the machine must be restarted after installing
        # Chocolatey.
        #
        if ($SCRIPT_DATA.WIN_VERSION -eq $WIN7SP1_VERSION_ID) {
            Restart-Computer
            PrepareFor-Exit -DownloadLocation $SCRIPT_DATA.DOWNLOAD_TO -PreviousSettingForEAP $PREVIOUS_EAP
            Exit
        }
    }

    if ($upgradeChocolatey) {
        Write-LogMessage "Ensuring latest version is installed..."
        choco upgrade -y chocolatey
        Write-LogMessage "...done!"
    }

    refreshenv

    ##### UPDATE .NET

    if ($SCRIPT_DATA.NET_VERSION -lt "4.5") {
        Write-LogMessage "Installing .NET 4.5..."
        choco install -y dotnet4.5
        Write-LogMessage "...done!"

        Restart-Computer
        PrepareFor-Exit -DownloadLocation $SCRIPT_DATA.DOWNLOAD_TO -PreviousSettingForEAP $PREVIOUS_EAP
        Exit
    }

    ##### UPDATE POWERSHELL

    if ($SCRIPT_DATA.PS_MAJOR_VERSION -lt 5) {
        Write-LogMessage "Ensuring latest version of PowerShell is installed..."
        choco install -y powershell
        Write-LogMessage "...done!"

        Restart-Computer
        PrepareFor-Exit -DownloadLocation $SCRIPT_DATA.DOWNLOAD_TO -PreviousSettingForEAP $PREVIOUS_EAP
        Exit
    }

    ##### CARBON

    Write-LogMessage "Ensuring PS Carbon is installed..."

    choco install -y carbon
    Import-Module "Carbon"

    Write-LogMessage "...done!"

    ##### READ LOCAL CONFIG

    Write-LogMessage "Reading local configuration file..."
    [Hashtable]$configData = Import-PowerShellDataFile (Join-Path $PSScriptRoot "config.psd1")
    Write-LogMessage "...done!"

    ##### CREATE LOCAL USER

    Write-LogMessage "Ensuring local user is present..."

    [SecureString]$passwordAsSecureString = ConvertTo-SecureString $configData.LocalAdmin.Password -AsPlainText -Force
    [PSCredential]$adminCredentials = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $configData.LocalAdmin.Username, $passwordAsSecureString

    Install-User `
        -Credential  $adminCredentials `
        -Description $configData.LocalAdmin.Description `
        -UserCannotChangePassword

    Write-LogMessage "...done!"

    Write-LogMessage "Ensuring local user is added to Administrators group..."

    Add-GroupMember `
        -Name Administrators `
        -Member $configData.LocalAdmin.Username

    Write-LogMessage "...done!"

    ##### ENABLE WINRM FOR ANSIBLE
    Write-LogMessage "Downloading script to enable WinRM for Ansible..."

    [String]$pathToAnsibleScript = (Join-Path $SCRIPT_DATA.DOWNLOAD_TO "__winerva__ConfigureRemotingForAnsible.ps1")
    Download-File -Url "https://raw.githubusercontent.com/ansible/ansible/devel/examples/scripts/ConfigureRemotingForAnsible.ps1" -Dest $pathToAnsibleScript

    Write-LogMessage "...done!"

    Write-LogMessage "Executing..."
    . $pathToAnsibleScript
    Write-LogMessage "...done!"

    ##### ADD HOSTNAME TO INVENTORY FILE

    Write-LogMessage "Checking if computer name is already included in inventory for Ansible..."

    try {
        $inventory = (Get-Content -Path $SCRIPT_DATA.INVENTORY_FILE)
        if ($inventory -eq $null) {
            $inventory = ""
        }
   }
    catch {
        $inventory = ""
    }

    if ($inventory.Contains($SCRIPT_DATA.HOSTNAME)) {
        Write-LogMessage "...present."
    }
    else {
        Write-LogMessage "not found, adding..."
        Add-Content -Path $SCRIPT_DATA.INVENTORY_FILE -Value "$([System.Environment]::NewLine)$($SCRIPT_DATA.HOSTNAME)"
        Write-LogMessage "...done!"
    }

    # The following was adapted from https://stackoverflow.com/a/11002660
    Write-LogMessage "Removing extraneous blank lines from inventory file..."
    (Get-Content $SCRIPT_DATA.INVENTORY_FILE) | ? { -not [System.String]::IsNullOrWhiteSpace($_) } | Set-Content $SCRIPT_DATA.INVENTORY_FILE
    Write-LogMessage "...done!"

    Write-LogMessage "+++++ INSTALLATION COMPLETE ++++"
}
catch {
    Write-LogMessage "****** ERROR ******"
    Write-LogMessage $_.Exception.Message
    Write-LogMessage $_.InvocationInfo.PositionMessage
    Write-LogMessage "*******************"
}
finally {
    PrepareFor-Exit -DownloadLocation $SCRIPT_DATA.DOWNLOAD_TO -PreviousSettingForEAP $PREVIOUS_EAP
}
